Mathlib/CategoryTheory/Limits/Shapes/BinaryProducts.lean

```diff
@@ -0,0 +1,1437 @@
+    [HasBinaryProduct Z W] [HasBinaryProduct Y W] :
+    prod.map (f ≫ g) (𝟙 W) = prod.map f (𝟙 W) ≫ prod.map g (𝟙 W) := by simp
+#align category_theory.limits.prod.map_comp_id CategoryTheory.Limits.prod.map_comp_id
+
+@[reassoc]
+theorem prod.map_id_comp {X Y Z W : C} (f : X ⟶ Y) (g : Y ⟶ Z) [HasBinaryProduct W X]
+    [HasBinaryProduct W Y] [HasBinaryProduct W Z] :
+    prod.map (𝟙 W) (f ≫ g) = prod.map (𝟙 W) f ≫ prod.map (𝟙 W) g := by simp
+#align category_theory.limits.prod.map_id_comp CategoryTheory.Limits.prod.map_id_comp
+
+/-- If the products `W ⨯ X` and `Y ⨯ Z` exist, then every pair of isomorphisms `f : W ≅ Y` and
+    `g : X ≅ Z` induces an isomorphism `prod.mapIso f g : W ⨯ X ≅ Y ⨯ Z`. -/
+@[simps]
+def prod.mapIso {W X Y Z : C} [HasBinaryProduct W X] [HasBinaryProduct Y Z] (f : W ≅ Y)
+    (g : X ≅ Z) : W ⨯ X ≅ Y ⨯ Z where
+  hom := prod.map f.hom g.hom
+  inv := prod.map f.inv g.inv
+#align category_theory.limits.prod.map_iso CategoryTheory.Limits.prod.mapIso
+
+instance isIso_prod {W X Y Z : C} [HasBinaryProduct W X] [HasBinaryProduct Y Z] (f : W ⟶ Y)
+    (g : X ⟶ Z) [IsIso f] [IsIso g] : IsIso (prod.map f g) :=
+  IsIso.of_iso (prod.mapIso (asIso f) (asIso g))
+#align category_theory.limits.is_iso_prod CategoryTheory.Limits.isIso_prod
+
+instance prod.map_mono {C : Type _} [Category C] {W X Y Z : C} (f : W ⟶ Y) (g : X ⟶ Z) [Mono f]
+    [Mono g] [HasBinaryProduct W X] [HasBinaryProduct Y Z] : Mono (prod.map f g) :=
+  ⟨fun i₁ i₂ h => by
+    apply prod.hom_ext
+    · rw [← cancel_mono f]
+      simpa using congr_arg (fun f => f ≫ prod.fst) h
+    · rw [← cancel_mono g]
+      simpa using congr_arg (fun f => f ≫ prod.snd) h⟩
+#align category_theory.limits.prod.map_mono CategoryTheory.Limits.prod.map_mono
+
+@[reassoc] -- Porting note: simp can prove these 
+theorem prod.diag_map {X Y : C} (f : X ⟶ Y) [HasBinaryProduct X X] [HasBinaryProduct Y Y] :
+    diag X ≫ prod.map f f = f ≫ diag Y := by simp
+#align category_theory.limits.prod.diag_map CategoryTheory.Limits.prod.diag_map
+
+@[reassoc] -- Porting note: simp can prove these
+theorem prod.diag_map_fst_snd {X Y : C} [HasBinaryProduct X Y] [HasBinaryProduct (X ⨯ Y) (X ⨯ Y)] :
+    diag (X ⨯ Y) ≫ prod.map prod.fst prod.snd = 𝟙 (X ⨯ Y) := by simp
+#align category_theory.limits.prod.diag_map_fst_snd CategoryTheory.Limits.prod.diag_map_fst_snd
+
+@[reassoc] -- Porting note: simp can prove these
+theorem prod.diag_map_fst_snd_comp [HasLimitsOfShape (Discrete WalkingPair) C] {X X' Y Y' : C}
+    (g : X ⟶ Y) (g' : X' ⟶ Y') :
+    diag (X ⨯ X') ≫ prod.map (prod.fst ≫ g) (prod.snd ≫ g') = prod.map g g' := by simp
+#align category_theory.limits.prod.diag_map_fst_snd_comp CategoryTheory.Limits.prod.diag_map_fst_snd_comp
+
+instance {X : C} [HasBinaryProduct X X] : IsSplitMono (diag X) :=
+  IsSplitMono.mk' { retraction := prod.fst }
+
+end ProdLemmas
+
+section CoprodLemmas
+
+-- @[reassoc (attr := simp)]
+@[simp] -- Porting note: removing reassoc tag since result is not hygenic (two h's) 
```

Comment:
Has this been reported as `reassoc` bug?
Is it a viable workaround to rename the `h` in the statement to `h'`?

Reply:
I started a [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/hygenic.20reassoc) topic but I couldn't find a prior mention.

Reply:
Yeah, I guess renaming it would be easier.